src/cmd/flux-kvs.c

```diff
@@ -816,6 +850,254 @@ int cmd_dir (optparse_t *p, int argc, char **argv)
+        return width;
+    if (ioctl (fd, TIOCGWINSZ, &w) == 0)
+        return w.ws_col;
+    if (optparse_hasopt (p, "columns"))
+        return 80;
+    return 1;
+}
+
+/* List the content of 'dir', arranging output in columns that fit 'win_width',
+ * and using a custom column width selected based on the longest entry name.
+ */
+static void list_kvs_dir_single (kvsdir_t *dir, int win_width, optparse_t *p)
+{
+    kvsitr_t *itr;
+    const char *name;
+    int col_width = get_dir_maxname (dir) + 4;
+    int col = 0;
+    char *namebuf;
+
+    if (!(namebuf = malloc (col_width + 2)))
+        log_err_exit ("malloc");
+    if (!(itr = kvsitr_create (dir)))
+        log_err_exit ("kvsitr_create");
+    while ((name = kvsitr_next (itr))) {
+        if (win_width != 0 && col + col_width > win_width && col > 0) {
+            printf ("\n");
+            col = 0;
+        }
+        strcpy (namebuf, name);
+        if (optparse_hasopt (p, "classify")) {
+            if (kvsdir_isdir (dir, name))
+                strcat (namebuf, ".");
+            else if (kvsdir_issymlink (dir, name))
+                strcat (namebuf, "@");
+        }
+        printf ("%-*s", min (col_width, win_width), namebuf);
+        col += col_width;
+    }
+    printf ("\n");
+    kvsitr_destroy (itr);
+    free (namebuf);
+}
+
+/* List contents of directory pointed to by 'key', descending into subdirs
+ * if -R was specified.  First the directory is listed, then its subdirs.
+ */
+static void list_kvs_dir (flux_t *h, const char *key, optparse_t *p,
+                          int win_width, bool print_label, bool print_vspace)
+{
+    flux_future_t *f;
+    kvsdir_t *dir;
+    kvsitr_t *itr;
+    const char *name, *json_str;
+
+    if (!(f = flux_kvs_lookup (h, FLUX_KVS_READDIR, key))
+                || flux_kvs_lookup_get (f, &json_str) < 0) {
+        log_err_exit ("%s", key);
+        goto done;
+    }
+
+    if (print_label)
+        printf ("%s%s:\n", print_vspace ? "\n" : "", key);
+    if (!(dir = kvsdir_create (h, NULL, key, json_str)))
+        log_err_exit ("kvsdir_create");
+    list_kvs_dir_single (dir, win_width, p);
+
+    if (optparse_hasopt (p, "recursive")) {
+        if (!(itr = kvsitr_create (dir)))
+            log_err_exit ("kvsitr_create");
+        while ((name = kvsitr_next (itr))) {
+            if (kvsdir_isdir (dir, name)) {
+                char *nkey;
+                if (!(nkey = kvsdir_key_at (dir, name))) {
+                    log_err ("%s: kvsdir_key_at failed", name);
+                    continue;
+                }
+                list_kvs_dir (h, nkey, p, win_width, print_label, true);
+                free (nkey);
+            }
+        }
+        kvsdir_destroy (dir);
+    }
+done:
+    flux_future_destroy (f);
+}
+
+/* List keys, arranging in columns so that they fit within 'win_width',
+ * and using a custom column width selected based on the longest entry name.
+ * The keys are popped off the list and freed we go.
```

Comment:
```freed we go```, just "freed"?  Or had another thought but didn't finish?

Reply:
Freed as we go, but I'll go with "freed".